### PR TITLE
[ML] Use perAllocation and perDeployment memory usage in the model assignment planner

### PR DESCRIPTION
**Pull request by valeriy42**
Building upon #98139, this PR extends the model assignment planning algorithms and the linear solver to use the extended memory fields. It also adds unit tests to verify the new behavior. 

I needed to adjust the old unit tests since we use the `estimateMemoryUsage` routine, which would compute `2*memoryBytes + 240 MB` as the memory requirement. Previously, in the unit tests, we were simply using `memoryBytes` field value.

### CODE REVIEW

File: docs/changelog/98874.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+pr: 98874
+summary: Use `perAllocation` and `perDeployment` memory usage in the model assignment
+  planner
```
Comment (droberts195) [inline docs/changelog/98874.yaml:3]:
```suggestion
summary: Estimate the memory required to deploy trained models more accurately
```

**Review by droberts195 — COMMENTED** [review]
Something that's really important to think about is what this will do in a mixed version cluster during an upgrade.

Does all the planner code run on the master node? If so then effectively the version of the master node determines whether we get this new planning behaviour or the old behaviour, and then the decisions get sent to the ML nodes.

But is there ever a situation where other nodes in the cluster get the size of models? Say when calculating ML memory stats in `NodeLoadDetector` or `MlAutoscalingResourceTracker`?

Also think about what would happen if the master node switched to being on a lower version during a rolling upgrade. Although no single node can ever get downgraded, it's possible for the master version to go backwards because there are multiple master-eligible nodes, and which one is elected could change while a rolling upgrade is part way through.

If it's too hard to reason about then one option might be to say we'll use the old logic until the entire cluster is upgraded to the newer version. However, I realise that in itself introduces complexity.

Please have a think through some of the possibilities and what changes could increase confidence that we won't get a load of users with failed deployments when they upgrade to the version with this change in.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlannerTests.java [hunk 2]
```diff
@@ -33,70 +33,207 @@
...
-            Node node = new Node("n_1", 1000, 8);
-            Deployment deployment = new Deployment("m_1", 1000, 8, 1, Map.of(), 0);
+            Node node = new Node("n_1", scaleNodeSize(100), 4);
+            Deployment deployment = new AssignmentPlan.Deployment("m_1", ByteSizeValue.ofMb(100).getBytes(), 1, 1, Map.of(), 0, 0, 0);
             AssignmentPlan plan = new AssignmentPlanner(List.of(node), List.of(deployment)).computePlan();
             assertModelFullyAssignedToNode(plan, deployment, node);
         }
+    }
+
+    public void testSingleModelThatFitsFullyOnSingleNode_NewMemoryFields() {
```
Comment (tveasey) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlannerTests.java:108]: I sort of feel like we could have a bit more testing of the new memory fields. One thing I think would be worthwhile as a minimum is a test which randomises over nodes memory, and model dynamic and static memory and checks that the final assignment doesn't have any memory violations.

**Review by tveasey — COMMENTED** [review]
I did a pass through. All the new logic looks correct to me, but I'd like to read through again tomorrow.

However, I think we're a bit light on testing for the new memory fields. Perhaps as a minimum a test that we don't get memory violations in random assignments which use non-zero static and dynamic model memory.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlannerTests.java (hunk 2 above)
Comment (tveasey) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AssignmentPlannerTests.java:108]: Looking more carefully I see there are more `_NewMemoryFields` tests below. But I do think randomized testing of the memory constraint is a good idea anyway.

**Comment by valeriy42** [comment]
@elasticmachine update branch

**Comment by valeriy42** [comment]
@elasticmachine update branch

**Comment by valeriy42** [comment]
Thank you for your review @droberts195 . I added the handling of the mixed cluster use case in [50d6d2e](https://github.com/elastic/elasticsearch/pull/98874/commits/50d6d2ea7eeebbb861b63e3c9504e22a75f90c5b). It would be great if you could take another look.

**Review by droberts195 — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
@valeriy42 this is failing CI because of this:

```
Forbidden method invocation: java.lang.String#format(java.lang.String,java.lang.Object[]) [Uses default locale]
in org.elasticsearch.upgrades.MlAssignmentPlannerUpgradeIT (MlAssignmentPlannerUpgradeIT.java:141)
```

You can fix that by using `Strings.format` instead of `String.format`. (Where `Strings` is an Elastic class that wraps up adding the locale.)

**Comment by valeriy42** [comment]
@elasticmachine update branch

**Review by tveasey — APPROVED** [review]
Good work on this! LGTM

**Pull request merged**